sklearn/_build_utils/cythonize.py

```diff
@@ -0,0 +1,192 @@
+            if r != 0:
+                raise Exception('Cython failed')
+    except OSError:
+        raise OSError('Cython needs to be installed')
+
+
+rules = {
+    '.pyx': process_pyx,
+}
+
+
+#
+# Hash db
+#
+def load_hashes(filename):
+    # Return { filename : (sha1 of input, sha1 of output) }
+    if os.path.isfile(filename):
+        hashes = {}
+        with open(filename, 'r') as f:
+            for line in f:
+                filename, inhash, outhash = line.split()
+                hashes[filename] = (inhash, outhash)
+    else:
+        hashes = {}
+    return hashes
+
+
+def save_hashes(hash_db, filename):
+    with open(filename, 'w') as f:
+        for key, value in sorted(hash_db.items()):
+            f.write("%s %s %s\n" % (key, value[0], value[1]))
+
+
+def sha1_of_file(filename):
+    h = hashlib.sha1()
+    with open(filename, "rb") as f:
+        h.update(f.read())
+    return h.hexdigest()
+
+
+#
+# Main program
+#
+
+def normpath(path):
+    path = path.replace(os.sep, '/')
+    if path.startswith('./'):
+        path = path[2:]
+    return path
+
+
+def get_hash(frompath, topath):
+    from_hash = sha1_of_file(frompath)
+    to_hash = sha1_of_file(topath) if os.path.exists(topath) else None
+    return (from_hash, to_hash)
+
+
+def process(path, fromfile, tofile, processor_function, hash_db):
+    fullfrompath = os.path.join(path, fromfile)
+    fulltopath = os.path.join(path, tofile)
+    current_hash = get_hash(fullfrompath, fulltopath)
+    if current_hash == hash_db.get(normpath(fullfrompath), None):
+        print('%s has not changed' % fullfrompath)
+        return
+
+    orig_cwd = os.getcwd()
+    try:
+        os.chdir(path)
+        print('Processing %s' % fullfrompath)
+        processor_function(fromfile, tofile)
+    finally:
+        os.chdir(orig_cwd)
+    # changed target file, recompute hash
+    current_hash = get_hash(fullfrompath, fulltopath)
+    # store hash in db
+    hash_db[normpath(fullfrompath)] = current_hash
+
+
+def find_process_files(root_dir):
+    print(root_dir)
+    hash_db = load_hashes(HASH_FILE)
+    for cur_dir, dirs, files in os.walk(root_dir):
+        for filename in files:
+            in_file = os.path.join(cur_dir, filename + ".in")
+            if filename.endswith('.pyx') and os.path.isfile(in_file):
+                continue
+            for fromext, function in rules.items():
+                if filename.endswith(fromext):
+                    toext = ".c"
+                    with open(os.path.join(cur_dir, filename), 'rb') as f:
+                        data = f.read()
+                        m = re.search(b"libcpp", data, re.I | re.M)
+                        if m:
+                            toext = ".cpp"
+                    fromfile = filename
+                    tofile = filename[:-len(fromext)] + toext
+                    process(cur_dir, fromfile, tofile, function, hash_db)
+                    save_hashes(hash_db, HASH_FILE)
```

Comment:
I believe it is meant for cythonization to be cached even when doing Ctrl-C. I guess it is not that useful though